src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/SignatureDecoder.cs

```diff
@@ -316,10 +322,27 @@ private TType DecodeTypeDefOrRef(ref BlobReader blobReader, SignatureTypeHandleC
                     }
                     return _provider.GetTypeFromReference(_metadataReaderOpt, typeRef, code);
 
+                case HandleKind.TypeSpecification:
+                    if (!allowSpec)
+                    {
+                        // To prevent cycles, the token following (CLASS | VALUETYPE) must not be a type spec.
+                        // https://github.com/dotnet/coreclr/blob/8ff2389204d7c41b17eff0e9536267aea8d6496f/src/md/compiler/mdvalidator.cpp#L6154-L6160
+                        throw new BadImageFormatException(SR.NotTypeDefOrRefHandle);
```

Comment:
@tmat I think in fact custom modifiers using type specs can cause cycles. because custom mods can contain type spec and type specs can contain custom mods. Will do some investigation.